### PR TITLE
Better HTML generation behaviour for Requirements_Backend

### PR DESCRIPTION
**Pull request by tractorcow**
Because string concatenation in Requirements_Backend is gross.

Merge with https://github.com/silverstripe/silverstripe-cms/pull/1793

### CODE REVIEW

**Comment by dhensby** [comment]
I don't like that we use `Form` to build HTML - can't we have an HTML builder or something?

**Comment by tractorcow** [comment]
Perhaps something is SilverStripe\View?

**Comment by dhensby** [comment]
> Perhaps something in SilverStripe\View?

or a library if there's a good one?

**Comment by sminnee** [comment]
My recommendation is that we push `create_tag()` to something in view, and have the form system call that instead of `FormField::create_tag()`. Form depending on View is better than the reverse.

**Comment by tractorcow** [comment]
I've moved create_tag to HTMLValue and made it non-static; This means that if we wanted to set a custom HTMLValue subclass, we could customise tag creation. :)

**Comment by tractorcow** [comment]
Note: CMS changes need merging too at https://github.com/silverstripe/silverstripe-cms/pull/1793, since this change is now breaking.

**Comment by dhensby** [comment]
looks much nicer

**Comment by tractorcow** [comment]
Is this mergable if we like the API?

**Comment by tractorcow** [comment]
Rebased.

**Comment by sminnee** [comment]
> I've moved create_tag to HTMLValue and made it non-static

I'm not convinced that the non-static API here is an improvement, as the method is essentially a stateless function.

If we do want to make it an instance method it should be a method of an `HTMLGenerator` service or something — it's not nothing to do with the responsibility of HTMLValue. At best, it a type of HTMLValueFactory.

File: src/Forms/GridField/GridField.php [hunk 1]
```diff
@@ -466,17 +467,17 @@ public function FieldHolder($properties = array())
         }
 
         // Display a message when the grid field is empty.
-
+        $htmlValue = HTMLValue::create();
```
Comment (flamerohr) [inline src/Forms/GridField/GridField.php]: would `$htmlValue = HTMLValue::singleton();` be better?

**Comment by flamerohr** [comment]
Was the concensus to move the `createTag` method to a factory or provide?

**Comment by dhensby** [comment]
I agree with Sam that the method and best as a static or on a service

**Comment by tractorcow** [comment]
Ok, so create new service, and make it static (similar to `Convert::raw2sql`)

**Comment by tractorcow** [comment]
Ok, done @dhensby :D Can you please review and make sure this new approach is better?

File: src/View/HTML.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    use Configurable;
+
+    /**
+     * List of HTML5 void elements
+     *
+     * @see https://www.w3.org/TR/html51/syntax.html#void-elements
+     * @config
+     * @var array
+     */
+    private static $void_elements = [
```
Comment (flamerohr) [inline src/View/HTML.php]: very nice, extendable by config :D

File: tests/php/View/HTMLTest.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    public function testVoidContentError()
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage("Void element \"link\" cannot have content");
+        $tag = HTML::createTag('link', [
+            'title' => 'HTML & Text',
+            'nullattr' => null,
+        ], 'Some <strong>content!</strong>');
+        $this->assertEquals('<a title="HTML &amp; Text">Some <strong>content!</strong></a>', $tag);
```
Comment (flamerohr) [inline tests/php/View/HTMLTest.php]: I'm confused with why it's asserting `'<a title="HTML &amp; Text">Some <strong>content!</strong></a>'`, does this assert even trigger?
if that's the case, perhaps add a comment above to state that.

Although I've noticed another `expectException()` below and doesn't have any asserts following the code which throws an exception

**Comment by tractorcow** [comment]
There was an issue where empty attributes were being generated where none should be (e.g. `disabled=""`) which broke behat.

**Comment by flamerohr** [comment]
I'll restart tests when the cms part is merged :)

**Pull request merged**